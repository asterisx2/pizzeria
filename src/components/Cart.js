import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import CartItem from './CartItem';

export default class Cart extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      total: 0,
      ordered: false
    };

    this.calculateTotal = this.calculateTotal.bind(this);
    this.removePizza = this.removePizza.bind(this);
    this.handleConfirm = this.handleConfirm.bind(this);
  }

  static propTypes = {
    handleRemovePizza: PropTypes.func.isRequired,
    cart: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        size: PropTypes.string.isRequired,
        quantity: PropTypes.number.isRequired,
        basePrice: PropTypes.number.isRequired,
        toppings: PropTypes.arrayOf(
          PropTypes.shape({
            __typename: PropTypes.string.isRequired,
            name: PropTypes.string.isRequired,
            price: PropTypes.number.isRequired
          }).isRequired
        ).isRequired
      }).isRequired
    ).isRequired
  };

  async removePizza(pizzaId) {
    await this.props.handleRemovePizza(pizzaId);
    this.calculateTotal();
  }

  handleConfirm() {
    this.removepizza();
    this.setState({ ordered: true });
  }

  calculateTotal() {
    let sum = 0;
    this.props.cart.forEach(pizza => {
      sum +=
        (pizza.basePrice +
          pizza.doughType.price +
          pizza.toppings.reduce((topSum, top) => (topSum += top.price), 0)) *
        pizza.quantity;
    });
    this.setState({
      total: sum
    });
  }

  componentDidMount() {
    this.calculateTotal();
  }

  removepizza() {
    this.props.cart.forEach(pizza => {
      this.props.handleRemovePizza(pizza.id);
    });
  }

  getImageSize(size) {
    switch (size) {
    case 'small':
      return '60px';
  
    case 'medium':
      return '70px';
  
    case 'large':
      return '80px';
  
    default:
      return '100px';
    }
  };

  render() {
    let pageContent;
    let footer;
    if (this.state.ordered) {
      pageContent = (
        <section>
          <div>
            <i 
              className="fa fa-check-circle shopping-cart-icon-large" 
              aria-hidden="true"/>
            <h1
              className="mt-2">
              Your order is on its way
            </h1>
          </div>
        </section>
      );
      footer = (
        <div 
          className="d-flex justify-content-center col">
          <Link 
            className="text-light" 
            to="/">
            <button 
              type="button" 
              className="btn btn-success">
              <i 
                className="fa fa-shopping-cart pr-2" 
                aria-hidden="true" />
              Order More Pizzas
            </button>
          </Link>
        </div>
      );
    } else if (this.props.cart.length === 0) {
      pageContent = (
        <section>
          <div>
            <i 
              className="fa fa-exclamation-circle shopping-cart-icon-large" 
              aria-hidden="true"></i>
            <h1
              className="mt-2">
              Your cart is empty
            </h1>
          </div>
        </section>
      );
      footer = (
        <div
          className="d-flex justify-content-center col">
          <Link 
            className="text-light" 
            to="/">
            <button
              type="button" 
              className="btn btn-success">
              <i 
                className="fa fa-shopping-cart pr-2" 
                aria-hidden="true" />
              Add Some Pizzas
            </button>
          </Link>
        </div>
      );
    } else {
      pageContent = (
        <div 
          className="fluid-width h-100">
          <h1>
            <i 
              className="fa fa-shopping-cart pr-2" 
              aria-hidden="true" />
            Your Cart
          </h1>
          <div>
            <div>
              {this.props.cart.map(pizza => {
                return (
                  <CartItem
                    key={pizza.id}
                    pizza={pizza}
                    removePizza={() => this.removePizza(pizza.id)}/>
                );
              })}
            </div>
            <div 
              className="w-100 d-flex justify-content-between font-medium mt-4">
              <span>
                Order Total: 
              </span>
              <span>
                &euro;
                {this.state.total.toFixed(2)}
              </span>
            </div>
          </div>
        </div>
      );
      footer = (
        <div 
          className="d-flex justify-content-between col">
          <Link 
            className="text-light" 
            to="/">
            <button 
              type="button" 
              className="btn btn-primary">
              <i 
                className="fa fa-shopping-cart pr-2" 
                aria-hidden="true" />
              Add More Pizzas
            </button>
          </Link>
          <button
            type="button"
            className="btn btn-success"
            onClick={this.handleConfirm}>
            <i className="fa fa-check pr-2" aria-hidden="true" />
            Confim Order
          </button>
        </div>
      );
    }
    return (
      <div 
        className="d-flex flex-column">
        <div 
          className="d-flex justify-content-between align-content-center p-3 theme-primary-color text-light">
          <div>
            <Link 
              className="text-light" 
              to='/'>
              <span 
                className="logo">
                PZ
              </span>
              <span className="pl-2 pr-2">
                | 
              </span>
              <span>
                Pizzeria
              </span>
            </Link>
          </div>
        </div>
        <div 
          className="flex-1 p-3 d-flex justify-content-center align-items-center overflow-y-auto">
          {pageContent}
        </div>
        <div 
          className="d-flex align-content-center p-3 theme-primary-color">
          {footer}
        </div>
      </div>
    );
  }
}
