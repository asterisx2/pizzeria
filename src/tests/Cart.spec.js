import React from 'react';
import Cart from '../components/Cart';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

describe('Cart Component', () => {
  it('renders without crashing', () => {
    let store = [];

    const tree = renderer
      .create(
        <MemoryRouter>
          <Cart cart={store} handleRemovePizza={() => { }} />
        </MemoryRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  // TODO
  it('handles remove pizza correctly', () => { });
});
