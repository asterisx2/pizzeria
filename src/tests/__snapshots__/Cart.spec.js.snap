// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Component renders without crashing 1`] = `
<div
  className="d-flex flex-column"
>
  <div
    className="d-flex justify-content-between align-content-center p-3 theme-primary-color text-light"
  >
    <div>
      <a
        className="text-light"
        href="/"
        onClick={[Function]}
      >
        <span
          className="logo"
        >
          PZ
        </span>
        <span
          className="pl-2 pr-2"
        >
          |
        </span>
        <span>
          Pizzeria
        </span>
      </a>
    </div>
  </div>
  <div
    className="flex-1 p-3 d-flex justify-content-center align-items-center overflow-y-auto"
  >
    <section>
      <div>
        <i
          aria-hidden="true"
          className="fa fa-exclamation-circle shopping-cart-icon-large"
        />
        <h1
          className="mt-2"
        >
          Your cart is empty
        </h1>
      </div>
    </section>
  </div>
  <div
    className="d-flex align-content-center p-3 theme-primary-color"
  >
    <div
      className="d-flex justify-content-center col"
    >
      <a
        className="text-light"
        href="/"
        onClick={[Function]}
      >
        <button
          className="btn btn-success"
          type="button"
        >
          <i
            aria-hidden="true"
            className="fa fa-shopping-cart pr-2"
          />
          Add Some Pizzas
        </button>
      </a>
    </div>
  </div>
</div>
`;
